File: 688.py
class Solution:
    def knightProbability(self, N: int, K: int, r: int, c: int) -> float:
        moveDirs = ((2, 1), (2, -1), (-2, 1), (-2, -1), (1, 2), (1, -2), (-1, 2), (-1, -2))
        dp = [[0] * N for _ in range(N)]
        dp[r][c] = 1.0
        print(dp)
        for step in range(K):
            dp2 = [[0] * N for _ in range(N)]
            for row in range(N):
                for col in range(N):
                    if dp[row][col] > 0:
                        for mr, mc in moveDirs:
                            newRow = row + mr
                            newCol = col + mc
                            if 0 <= newRow < N and 0 <= newCol < N:
                                dp2[newRow][newCol] += dp[row][col] / 8.0
            dp = dp2
            print(dp)
        #print(dp)
        return sum(map(sum, dp))

if __name__ == '__main__':
    s = Solution()
    print(s.knightProbability(3,2,0,0))

File: 688.go
package leetcode

func knightProbability(n int, k int, row int, column int) float64 {
	dp := make([][][]float64, k+1)
	dirs := []struct{ x, y int }{
		{1, 2},
		{1, -2},
		{-1, 2},
		{-1, -2},
		{2, 1},
		{2, -1},
		{-2, 1},
		{-2, -1},
	}
	for step := 0; step <= k; step++ {
		dp[step] = make([][]float64, n)
		for i := range dp[step] {
			dp[step][i] = make([]float64, n)
			for j := range dp[step][i] {
				if step == 0 {
					dp[step][i][j] = 1.0
				} else {
					for _, d := range dirs {
						x := d.x + i
						y := d.y + j
						if 0 <= x && x < n && 0 <= y && y < n {
							dp[step][i][j] += dp[step-1][x][y] / 8.0
						}
					}
				}
			}
		}
	}
	return dp[k][row][column]
}
